### PR TITLE
sshguard + service: init at 2.0.0

### PR DESCRIPTION
**Pull request by sargon**
###### Motivation for this change

sshguard has a smaller footprint than e.g. fail2ban and support blacklisting repeating attackers.

###### Things done

- [x] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [x] Linux
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: pkgs/tools/security/sshguard/default.nix [hunk 1]
```diff
@@ -0,0 +1,41 @@
+{ stdenv, fetchgit, autoreconfHook, yacc, flex, pkgs}:
```
Comment (c0bw3b) [inline pkgs/tools/security/sshguard/default.nix]: You don't need to pass the full `pkgs` to your derivation

File: pkgs/tools/security/sshguard/default.nix [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+  meta = with stdenv.lib; {
+    description = "SSHGuard protects hosts from brute-force attacks";
+    longDescription = ''
+      SSHGuard can read log messages from various input sources. Log messages are parsed, line-by-line, for recognized patterns.
+      If an attack, such as several login failures within a few seconds, is detected, the offending IP is blocked.
+    '';
+    homepage = https://sshguard.net;
+    license = stdenv.lib.licenses.bsd3;
+    maintainers = [ maintainers.sargon ];
+    platforms = stdenv.lib.platforms.all;
```
Comment (c0bw3b) [inline pkgs/tools/security/sshguard/default.nix]: You don't need to rewrite `stdenv.lib` for licence and platforms.
Also on platforms : you are explicitly building for `iptables` backend so it won't work on Darwin.

File: pkgs/tools/security/sshguard/default.nix [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+stdenv.mkDerivation rec {
+  version = "1.7.1";
+  name = "sshguard-${version}";
+
+  src = fetchgit {
+    url = "https://bitbucket.org/sshguard/sshguard.git";
+    rev = "8600fd5b70cd80fad87c414fd7c410082991a209";
+    sha256 = "1i2ic8nca4f61pl2ggjp4x4an37hhvjgk2wv4smjsj6lvz3bw06a";
+  };
```
Comment (c0bw3b) [inline pkgs/tools/security/sshguard/default.nix]: Why using `fetchgit` when you can get the stable release source from their SourceForge repo using `fetchurl` ?
https://sourceforge.net/projects/sshguard/files/sshguard/

**Review by c0bw3b — CHANGES_REQUESTED** [review]

**Comment by c0bw3b** [comment]
Ah nice!
I also prefer `sshguard` over `fail2ban` since it doesn't need a full Python stack. It also supports IPv6, unlike f2b.

I've made some comments on your `sshguard` derivation mainly. The service module looks good.
You should also squash your 3 commits into one, please.

**Comment by sargon** [comment]
Thanks for the review, I fixed most of your suggestions, but struggling with the multi platform support. 
When I have a nice solution for that, I can fix it all up into one patch.

**Comment by sargon** [comment]
Okay. I didn't see to get it right. So I reduced it to linux support and fix it all up into one patch. So someone with more experience in nix packages can pick it up. Or I will do it later when I got my mind around the builtins and stuff...

**Comment by c0bw3b** [comment]
It's the safe course to target the package at Linux platforms only. It could still be enhanced for other platforms later on.

Now on your `sshguard` derivation : you started your package with 1.7.1 version in mind but you are now fetching the source of the newly released 2.0.0 :)

Either you stick to v1.7.1 and your fetching should be something like 
```Nix
src = fetchurl {
  url = "mirror://sourceforge/sshguard/sshguard-${version}.tar.gz";
  sha256 = "...";
}
```

... or you try to build the v2.0.0 [[release note](https://www.sshguard.net/litenewz/feeds/14)]
1. you wouldn't need the `configureFlags` anymore because v2.0.0 "_build and install all backend scripts by default_"
2. but v2.0.0 also "_requires a configuration file to start_" so it would force you to rework your service module

**Comment by sargon** [comment]
Okay. You are right. I picked the 2.0.0 in advance to your urlfetch suggestion, but to my shame I didn't study the release notes that intensive. I will further adopt your proposals. Should we, for the sake of the PR title close this one and start a new one stating the version number?

**Comment by fpletz** [comment]
We can change the title, no problem. Just force-push your changes.

The code looks also good to me apart from @c0bw3b's comments. If you're ready I can give this a spin and merge.

**Comment by sargon** [comment]
Okay, this should do the job.

**Comment by c0bw3b** [comment]
Tested here with `nox-review` : it builds and runs fine.

**Comment by sargon** [comment]
@fpletz: I think you can give it a spin and merge, if you are satisfied with the changes.

**Comment by sargon** [comment]
Another fixup from my side. 

Leaving creation of the ipsets to sshguards fw scripts seems no good idea. Sometimes the post startup actions gets started so fast, on some of my systems, that the first run of the sshguard firewall scripts hasn't happened. Hence the startup failed and systemd is immediately restarting the service. So we run into a classic timing problem here. So now the problematic system is starting the service correctly during boot. @fpletz merge?

**Comment by fpletz** [comment]
@sargon You have pushed the wrong commit. This seems to be a module for nullmailer. :)

**Comment by sargon** [comment]
Narf. Must have happened this morning, just wanted to have that online. Fixed the branch, next time I should really not take the master branch for that.

**Comment by sargon** [comment]
@fpletz may you have another look?

**Pull request merged**